public/app/features/trails/TrailStore.ts

```diff
@@ -0,0 +1,129 @@
+import { debounce } from 'lodash';
+import { v4 as uuidv4 } from 'uuid';
+
+import { SceneObject, SceneObjectRef, SceneObjectUrlValues, getUrlSyncManager, sceneUtils } from '@grafana/scenes';
+
+import { DataTrail } from './DataTrail';
+import { TrailStepType } from './DataTrailsHistory';
+import { BOOKMARKED_TRAILS_KEY, RECENT_TRAILS_KEY } from './shared';
+
+const MAX_RECENT_TRAILS = 20;
+
+export interface SerializedTrail {
+  key: string;
+  history: Array<{
+    urlValues: SceneObjectUrlValues;
+    type: TrailStepType;
+    description: string;
+  }>;
+}
+
+export class TrailStore {
+  private _recent: Array<SceneObjectRef<DataTrail>> = [];
+  private _bookmarks: Array<SceneObjectRef<DataTrail>> = [];
+  private _save;
+
+  constructor() {
+    this._recent = this._loadFromStorage(RECENT_TRAILS_KEY);
+    this._bookmarks = this._loadFromStorage(BOOKMARKED_TRAILS_KEY);
+
+    this._save = debounce(() => {
+      const serializedRecent = this._recent
+        .slice(0, MAX_RECENT_TRAILS)
+        .map((trail) => this._serializeTrail(trail.resolve()));
+      localStorage.setItem(RECENT_TRAILS_KEY, JSON.stringify(serializedRecent));
+
+      const serializedBookmarks = this._bookmarks.map((trail) => this._serializeTrail(trail.resolve()));
+      localStorage.setItem(BOOKMARKED_TRAILS_KEY, JSON.stringify(serializedBookmarks));
+    }, 1000);
+  }
+
+  private _loadFromStorage(key: string) {
+    const list: Array<SceneObjectRef<DataTrail>> = [];
+    const storageItem = localStorage.getItem(key);
+
+    if (storageItem) {
+      const serializedTrails: SerializedTrail[] = JSON.parse(storageItem);
+      for (const t of serializedTrails) {
+        const trail = this._deserializeTrail(t);
+        list.push(trail.getRef());
```

Comment:
Are we deserializing all of them just to get the ref?
Is there some other deserializing going on besides this `_deserializeTrail` method, or does it have some side-effects where it stores them somewhere?

Reply:
Ah, it looks like the `getRef()` is more than just a simple key.
`trail.resolve()` seems to help us get the trail back where it is needed.
Scenes magic?

Reply:
Interesting. Thanks.